module/executiondatasync/optimistic_syncing/execution_result_query_provider_test.go

```diff
@@ -0,0 +1,298 @@
+			r.ExecutorID = allExecutionNodes[i].NodeID
+			r.ExecutionResult = *executionResult
+			receipts[i] = r
+		}
+
+		suite.receipts.On("ByBlockID", block.ID()).Return(receipts, nil)
+		suite.setupIdentitiesMock(allExecutionNodes)
+
+		requiredExecutors := allExecutionNodes[0:2].NodeIDs()
+		criteria := Criteria{
+			AgreeingExecutors: 1,
+			RequiredExecutors: requiredExecutors,
+			ResultInFork:      executionResult.ID(),
+		}
+
+		query, err := suite.provider.ExecutionResultQuery(block.ID(), criteria)
+		suite.Require().NoError(err)
+
+		suite.Assert().Equal(criteria.ResultInFork, query.ExecutionResult.ID())
+	})
+
+	suite.Run("successful query with different block results", func() {
+		otherResult := unittest.ExecutionResultFixture()
+		// Create 3 receipts with the same result (executionResult) and 2 with a different result (otherResult)
+		receipts := make(flow.ExecutionReceiptList, totalReceipts)
+		for i := 0; i < 3; i++ {
+			r := unittest.ReceiptForBlockFixture(&block)
+			r.ExecutorID = allExecutionNodes[i].NodeID
+			r.ExecutionResult = *executionResult
+			receipts[i] = r
+		}
+		for i := 3; i < totalReceipts; i++ {
+			r := unittest.ReceiptForBlockFixture(&block)
+			r.ExecutorID = allExecutionNodes[i].NodeID
+			r.ExecutionResult = *otherResult
+			receipts[i] = r
+		}
+
+		suite.receipts.On("ByBlockID", block.ID()).Return(receipts, nil)
+		suite.setupIdentitiesMock(allExecutionNodes)
+
+		requiredExecutors := allExecutionNodes[0:3].NodeIDs()
+		criteria := Criteria{
+			AgreeingExecutors: 2,
+			RequiredExecutors: requiredExecutors,
+		}
+
+		query, err := suite.provider.ExecutionResultQuery(block.ID(), criteria)
+		suite.Require().NoError(err)
+
+		suite.Require().Equal(executionResult.ID(), query.ExecutionResult.ID())
+		suite.Assert().ElementsMatch(requiredExecutors, query.ExecutionNodes.NodeIDs())
+	})
+
+	suite.Run("insufficient agreeing executors returns error", func() {
+		// Create a fresh block for this test to ensure proper isolation
+		insufficientBlock := unittest.BlockFixture()
+
+		// Create a scenario where we have receipts but no result has enough agreeing executors
+
+		// Create only 1 receipt with 1 execution result
+		r := unittest.ReceiptForBlockFixture(&insufficientBlock)
+		r.ExecutorID = allExecutionNodes[0].NodeID
+		r.ExecutionResult = *unittest.ExecutionResultFixture()
+		receipts := flow.ExecutionReceiptList{
+			r,
+		}
+
+		// Set up a separate mock call for this specific block
+		suite.receipts.On("ByBlockID", insufficientBlock.ID()).Return(receipts, nil).Once()
+
+		requiredExecutors := allExecutionNodes[0:1].NodeIDs()
+		criteria := Criteria{
+			AgreeingExecutors: 2,
+			RequiredExecutors: requiredExecutors,
+		}
+
+		_, err := suite.provider.ExecutionResultQuery(insufficientBlock.ID(), criteria)
+		suite.Require().Error(err)
+
+		suite.Assert().True(backend.IsInsufficientExecutionReceipts(err))
+	})
+
+	suite.Run("required executors not found returns error", func() {
```

Comment:
can you add tests covering different scenarios where the user's criteria overrides the operators, and when operator defaults are used